building/tooling/test-generators.md

```diff
@@ -1,134 +1,256 @@
+Not all exercises have a `canonical-data.json` file.
+If case they don't, you'll need to manually create the tests, as there is no data for the Test Generator to work with.
+```
+
+#### Structure
+
+Canonical data is defined in a JSON object.
+This object contains a `"cases"` field which contains the test cases.
+These test cases (normally) correspond one-to-one to tests in your track.
+
+Each test case has a couple of properties, with the description, property, input value(s) and expected value being the most important ones.
+Here is a (partial) example of the [canonical-data.json file of the leap exercise](https://github.com/exercism/problem-specifications/blob/main/exercises/leap/canonical-data.json):
+
+```json
+{
+  "exercise": "leap",
+  "cases": [
+    {
+      "uuid": "6466b30d-519c-438e-935d-388224ab5223",
+      "description": "year not divisible by 4 in common year",
+      "property": "leapYear",
+      "input": {
+        "year": 2015
+      },
+      "expected": false
+    },
+    {
+      "uuid": "4fe9b84c-8e65-489e-970b-856d60b8b78e",
+      "description": "year divisible by 4, not divisible by 100 in leap year",
+      "property": "leapYear",
+      "input": {
+        "year": 1996
+      },
+      "expected": true
+    }
+  ]
+}
+```
+
+The Test Generator's main responsibility is to transform this JSON data into track-specific tests.
+Here's how the above JSON could translate into Nim test code:
 
-- [Creating tests from scratch](#creating-tests-from-scratch)
-- [Reproducing tests for updates](#reproducing-tests-for-updates)
+```nim
+import unittest
+import leap
 
-The reason for this distinction is "designing the exercise" vs. "production-ready code".
+suite "Leap":
+  test "year not divisible by 4 in common year":
+    check isLeapYear(2015) == false
 
-When creating tests from scratch the test generator should provide all the information contained in `canonical-data.json` in the resulting files.
-This enables contributors to simply open up the generated test file(s) and find all relevant information interwoven with the tracks boilerplate code.
-They then design the exercise's tests and student facing code based on these files rather than on the original `canonical-data.json`.
-As there is no knowledge of exercise specific things, yet, a one-fits-all template targeting the boilerplate code can be used.
+  test "year divisible by 4, not divisible by 100 in leap year":
+    check isLeapYear(1996) == true
+```
+
+The structure of the `canonical-data.json` file is [well documented](https://github.com/exercism/problem-specifications?tab=readme-ov-file#test-data-canonical-datajson) and it also has a [JSON schema](https://github.com/exercism/problem-specifications/blob/main/canonical-data.schema.json) definition.
+
+##### Nesting
+
+Some exercises use nesting, where `cases` are nested in other `cases` keys.
+Only the innermost (leaf) `cases` will actually have any test cases, their parent `cases` will only ever be used for grouping.
```

Comment:
This is not the case. It may have been a design goal, but it isn't in reality. See e.g. https://github.com/exercism/problem-specifications/blob/main/exercises/binary-search-tree/canonical-data.json

Reply:
I stand corrected! I've updated.